File: merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    struct compare{
        bool operator()(ListNode* a,ListNode* b){
            return a->val > b->val;
        }
    };
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*,vector<ListNode*>,compare> heap;
        for(auto i:lists){if(i!=NULL){heap.push(i);}}
        ListNode* ll = new ListNode(0);
        ListNode* ret = ll;
        while(heap.size()>0){
            ListNode* t = heap.top();heap.pop();
            ll->next = new ListNode(t->val);
            ll = ll->next;
            t = t->next;
            if(t!=NULL){
                heap.push(t);
            }
        }
        return ret->next;
    }
};

File: merge-k-sorted-lists/merge-k-sorted-lists.java
// *
//  * Definition for singly-linked list.
//  * public class ListNode {
//  *     int val;
//  *     ListNode next;
//  *     ListNode() {}
//  *     ListNode(int val) { this.val = val; }
//  *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
//  * }

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> pq = new PriorityQueue<>(new Comparator<ListNode>(){
            public int compare(ListNode o1,ListNode o2){
                return o1.val - o2.val;
            }
        });
        for(ListNode i:lists){if(i!=null){pq.add(i);}}
        ListNode ans = new ListNode(1);
        ListNode sol = ans;
        while(pq.size()>0){
            ListNode top = pq.poll();
            ans.next = new ListNode(top.val);
            ans = ans.next;
            top = top.next;
            if(top != null){pq.add(top);}
        }
        return sol.next;
    }
}